packages/aws-cdk/lib/util/asset-publishing.ts

```diff
@@ -215,4 +217,48 @@ class PublishingProgressListener implements cdk_assets.IPublishProgressListener
     const handler = this.quiet && type !== 'fail' ? debug : EVENT_TO_LOGGER[type];
     handler(`[${event.percentComplete}%] ${type}: ${event.message}`);
   }
+}
+
+const PUBLISH_ASSET_PROOF_SYM = Symbol('publish_assets_proof');
+
+/**
+ * This interface represents proof that assets have been published.
+ *
+ * Objects of this type can only be obtained by calling functions in this module.
+ */
+export interface AssetsPublishedProof {
+  [PUBLISH_ASSET_PROOF_SYM]: true;
+}
+
+/**
+ * If you promise that you will publish assets for every element of an array, this function will attest it for every size of an array.
+ *
+ * Without this function, there would be no way to get `AssetsPublishedProof` for an empty array,
+ * since you'd have no single element to call `publishAssets` on.
+ */
+export async function multipleAssetPublishedProof<A>(xs: A[], block: (x: A) => Promise<AssetsPublishedProof>): Promise<AssetsPublishedProof> {
+  for (const x of xs) {
+    // Don't even care about the return value, just the type checking is good enough.
+    await block(x);
+  }
+  return { [PUBLISH_ASSET_PROOF_SYM]: true };
+}
+
+/**
+ * Conjure an `AssetsPublishedProof` out of thin air.
+ *
+ * This is only allowed in one location: the function `deployStack` in
+ * `deploy-stack.ts`.  Do not call this function anywhere but there.
```

Comment:
Meh. So this is the Get-out-of-jail-free card if you pinky promise to not misuse it?

Reply:
I mean. Yes. It's a pragmatic compromise without completely overhauling the codebase. The name should be clear enough to be flagged in code reviews if used for any other purpose.